analyzers/src/SonarAnalyzer.Common/Rules/Utilities/UtilityAnalyzerBase.cs

```diff
@@ -82,45 +83,62 @@ public abstract class UtilityAnalyzerBase<TSyntaxKind, TMessage> : UtilityAnalyz
 
         protected virtual bool AnalyzeUnchangedFiles => false;
 
-        protected virtual IEnumerable<TMessage> CreateAnalysisMessages(CompilationAnalysisContext c) => Enumerable.Empty<TMessage>();
+        protected virtual IEnumerable<TMessage> CreateAnalysisMessages(Compilation compilation) => Enumerable.Empty<TMessage>();
 
         protected UtilityAnalyzerBase(string diagnosticId, string title) : base(diagnosticId, title) { }
 
         protected sealed override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterCompilationAction(c =>
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                ReadParameters(context, startContext.Options, startContext.Compilation);
+                if (!IsAnalyzerEnabled)
+                {
+                    return;
+                }
+
+                ConcurrentStack<TMessage> treeMessages = new();
+                SonarAnalysisContext.TryGetValueDelegate<ProjectConfigReader> tryGetValue = startContext.TryGetValue;
+
+                startContext.RegisterSemanticModelAction(modelContext =>
                 {
-                    ReadParameters(context, c);
-                    if (!IsAnalyzerEnabled)
+                    var semanticModel = modelContext.SemanticModel;
+                    var syntaxTree = semanticModel.SyntaxTree;
```

Comment:
Can be simplified in a new code
```suggestion
                    var tree = semanticModel.SyntaxTree;
```